### PR TITLE
NodeMaterial: serialize/deserialize and Material.fromType()

### PR DESCRIPTION
**Pull request by sunag**
Related issue: https://github.com/mrdoob/three.js/pull/22552

# Introduction

This PR add support to serialize `NodeMaterial` in the `Scene Graph` of `Three.js`.

### Node.serialize() / Node.deserialize()
This can be used to `serialize` / `deserialize` specific `Node` fields, these methods can be extended.

### Node.fromType()
It can be used to create extended `Nodes` and share the same `NodeMaterialLoader`.

### Loaders

`NodeLoader` - only Nodes
`NodeMaterialLoader` - Nodes in Materials
`NodeObjectLoader` - Nodes in Scene Graph

### Serialization test 
- `webgl_materials_standard_nodes` - serialize objects after load to test

# Core

### Object3D
- `src/core/Object3D.js` - add `nodes` to `toJSON()`

### Material.fromType()
- `src/loaders/MaterialLoader.js`
- `src/materials/Material.js`
- `src/materials/Materials.js`
- Commit - https://github.com/mrdoob/three.js/commit/4cdbfac27f407c512a9f8e147ed792a65d1d1170

This contribution is funded by [Google via Igalia](https://igalia.com).

### CODE REVIEW

**Comment by mrdoob** [comment]
Hmm, one problem with this approach is that it makes `NodeMaterial` non-treeshakeable 😕

I was hoping that `NodeMaterial` was going to bring us closer to smaller builds, but I suspect this would actually produce bigger builds?

**Comment by mrdoob** [comment]
One of the key things about three.js is the fact that [it's ~150kb gzipped](https://bundlephobia.com/package/three@0.136.0).

Could you do some tests to see how the new `NodeMaterial` system impacts on this? For example, what would be the bundle size of a simple example that displays a red cube with the `WebGPURenderer`+`NodeMaterial`?

**Comment by sunag** [comment]
> Hmm, one problem with this approach is that it makes NodeMaterial non-treeshakeable 😕

I think you are referring to `Node.fromType()`?
In this case, we can just move NodeLib of Loader to a variable, the same that already happened with the textures in `MaterialLoader.setTextures()`.

Because some others approaches like `Node.serialize()` I was doing it tree shakeable.

**Comment by sunag** [comment]
I build here a .min files to reference in `gzipped`:

`145 KB` - `Three.js `
```js
import * as THREE from './three.min.js';
```

`152 KB` - `Three.js` + `WebGLNodeBuilder` + `ColorNode `
```js
import * as THREE from './three.min.js';
import ColorNode from './jsm/renderers/nodes/inputs/ColorNode.js';
import { nodeFrame } from './jsm/renderers/webgl/nodes/WebGLNodes.js';
```

`155 KB` - `Three.js` + `WebGLNodeBuilder` + `All Nodes`
```js
import * as THREE from './three.min.js';
import * as Nodes from './jsm/renderers/nodes/Nodes.js';
import { nodeFrame } from './jsm/renderers/webgl/nodes/WebGLNodes.js';
```

**Comment by mrdoob** [comment]
Nice!

Just remembered that we have a script that helps visualizing tree-shaking: https://github.com/mrdoob/three.js/pull/21657

File: examples/jsm/renderers/nodes/materials/NodeMaterial.js [hunk 1]
```diff
@@ -0,0 +1,98 @@
+import { Material, ShaderMaterial } from '../../../../../build/three.module.js';
```
Comment (mrdoob) [inline examples/jsm/renderers/nodes/materials/NodeMaterial.js:1]: I think this should be `import { Material, ShaderMaterial } from 'three'` now.

**Comment by mrdoob** [comment]
Thanks!

**Comment by sunag** [comment]
I was thinking of an approach to register classes imported. Like:

```js
class NoiseNode extends Node {
...
}

Node.library.add( NodeNode );
```

So we can register only the imported classes and preserve this design.

**Pull request merged**